File: src/adapters/anthropic/from-ir.ts
import type {
  IrContent,
  IrMessage,
  IrRequest,
  IrTextContent,
  IrToolResultContent,
  IrToolUseContent,
} from '../../ir/types.js';
import type { AnthropicRequestBody } from './to-ir.js';
import { applyPassthrough, assignIfPresent } from './passthrough.js';

function fromTextContent(block: IrTextContent): Record<string, unknown> {
  const wire: Record<string, unknown> = { type: 'text', text: block.text };
  assignIfPresent(wire, 'cache_control', block.cacheControl);
  return applyPassthrough(wire, block.passthrough);
}

function fromToolUseContent(block: IrToolUseContent): Record<string, unknown> {
  const wire: Record<string, unknown> = {
    type: 'tool_use',
    id: block.id,
    name: block.name,
    input: block.input,
  };
  assignIfPresent(wire, 'cache_control', block.cacheControl);
  return applyPassthrough(wire, block.passthrough);
}

function fromToolResultContent(block: IrToolResultContent): Record<string, unknown> {
  const wire: Record<string, unknown> = {
    type: 'tool_result',
    tool_use_id: block.toolUseId,
  };
  assignIfPresent(
    wire,
    'content',
    fromContentField(block.content, block.isContentString),
  );
  assignIfPresent(wire, 'is_error', block.isError);
  assignIfPresent(wire, 'cache_control', block.cacheControl);
  return applyPassthrough(wire, block.passthrough);
}

function fromContentBlock(block: IrContent): unknown {
  if (block.kind === 'text') return fromTextContent(block);
  if (block.kind === 'tool_use') return fromToolUseContent(block);
  if (block.kind === 'tool_result') return fromToolResultContent(block);
  return block.raw;
}

/**
 * Reproduces the string-or-array form the content arrived in. A string form is
 * only recoverable from a single text block, which is what `toIR` produced.
 */
function fromContentField(content: IrContent[], isContentString?: boolean): unknown {
  if (isContentString !== true)
    return content.length > 0 ? content.map(fromContentBlock) : undefined;
  const first = content[0];
  if (first !== undefined && first.kind === 'text') return first.text;
  return content.map(fromContentBlock);
}

function fromMessage(message: IrMessage): Record<string, unknown> {
  const wire: Record<string, unknown> = { role: message.role };
  const content = fromContentField(message.content, message.isContentString);
  wire['content'] = content === undefined ? [] : content;
  return applyPassthrough(wire, message.passthrough);
}

function fromSystem(request: IrRequest): unknown {
  const { system } = request;
  if (system === null) return undefined;
  if (request.extensions.isSystemString === true) {
    const first = system[0];
    if (first !== undefined && first.kind === 'text') return first.text;
  }
  return system.map(fromContentBlock);
}

/**
 * Converts the IR back to an Anthropic request body. Passthrough fields are
 * restored last so a modelled key never shadows the value that arrived.
 */
export function fromIR(request: IrRequest): AnthropicRequestBody {
  const body: Record<string, unknown> = {
    model: request.model,
    max_tokens: request.maxTokens,
  };
  assignIfPresent(body, 'system', fromSystem(request));
  body['messages'] = request.messages.map(fromMessage);
  if (request.tools.length > 0) body['tools'] = request.tools.map((tool) => tool.raw);
  return applyPassthrough(body, request.passthrough);
}

File: internal/adapters/anthropic/fromir.go
package anthropic

import "github.com/carlelieser/caveman/internal/ir"

func fromTextContent(block *ir.TextContent) ir.Value {
	wire := ir.NewObject()
	wire.Set("type", ir.String("text"))
	wire.Set("text", ir.String(block.Text))
	setIfPresent(wire, "cache_control", block.CacheControl)
	return inKeyOrder(applyPassthrough(wire, block.Passthrough), block.KeyOrder)
}

func fromToolUseContent(block *ir.ToolUseContent) ir.Value {
	wire := ir.NewObject()
	wire.Set("type", ir.String("tool_use"))
	wire.Set("id", ir.String(block.ID))
	wire.Set("name", ir.String(block.Name))
	if block.HasInput {
		wire.Set("input", block.Input)
	}
	setIfPresent(wire, "cache_control", block.CacheControl)
	return inKeyOrder(applyPassthrough(wire, block.Passthrough), block.KeyOrder)
}

func fromToolResultContent(block *ir.ToolResultContent) ir.Value {
	wire := ir.NewObject()
	wire.Set("type", ir.String("tool_result"))
	wire.Set("tool_use_id", ir.String(block.ToolUseID))
	setIfPresent(wire, "content", fromContentField(block.Content, block.IsContentString))
	if block.IsError != nil {
		wire.Set("is_error", ir.Bool(*block.IsError))
	}
	setIfPresent(wire, "cache_control", block.CacheControl)
	return inKeyOrder(applyPassthrough(wire, block.Passthrough), block.KeyOrder)
}

func fromContentBlock(block ir.Content) ir.Value {
	switch typed := block.(type) {
	case *ir.TextContent:
		return fromTextContent(typed)
	case *ir.ToolUseContent:
		return fromToolUseContent(typed)
	case *ir.ToolResultContent:
		return fromToolResultContent(typed)
	case *ir.ThinkingContent:
		return typed.Raw
	case *ir.OpaqueContent:
		return typed.Raw
	}
	return ir.Null{}
}

func fromContentBlocks(content []ir.Content) ir.Array {
	items := make(ir.Array, len(content))
	for i, block := range content {
		items[i] = fromContentBlock(block)
	}
	return items
}

// fromContentField reproduces the string-or-array form the content arrived in.
// A string form is only recoverable from a single text block, which is what
// ToIR produced. A nil result means the field is absent.
func fromContentField(content []ir.Content, isContentString bool) ir.Value {
	if !isContentString {
		if len(content) == 0 {
			return nil
		}
		return fromContentBlocks(content)
	}
	if len(content) > 0 {
		if text, ok := content[0].(*ir.TextContent); ok {
			return ir.String(text.Text)
		}
	}
	return fromContentBlocks(content)
}

func fromMessage(message ir.Message) ir.Value {
	wire := ir.NewObject()
	// An empty role is the absent one: the IR types role as a string, so a
	// message that arrived without one must not gain an empty one here.
	if message.Role != "" {
		wire.Set("role", ir.String(message.Role))
	}
	content := fromContentField(message.Content, message.IsContentString)
	if content == nil {
		content = ir.Array{}
	}
	wire.Set("content", content)
	return inKeyOrder(applyPassthrough(wire, message.Passthrough), message.KeyOrder)
}

func fromSystem(request ir.Request) ir.Value {
	if !request.HasSystem {
		return nil
	}
	if request.Extensions.IsSystemString && len(request.System) > 0 {
		if text, ok := request.System[0].(*ir.TextContent); ok {
			return ir.String(text.Text)
		}
	}
	return fromContentBlocks(request.System)
}

// FromIR converts the IR back to an Anthropic request body. Passthrough fields
// are restored last so a modelled key never shadows the value that arrived.
func FromIR(request ir.Request) *ir.Object {
	body := ir.NewObject()
	body.Set("model", ir.String(request.Model))
	body.Set("max_tokens", request.MaxTokens)
	setIfPresent(body, "system", fromSystem(request))

	messages := make(ir.Array, len(request.Messages))
	for i, message := range request.Messages {
		messages[i] = fromMessage(message)
	}
	body.Set("messages", messages)

	if len(request.Tools) > 0 {
		tools := make(ir.Array, len(request.Tools))
		for i, tool := range request.Tools {
			tools[i] = tool.Raw
		}
		body.Set("tools", tools)
	}

	return inKeyOrder(applyPassthrough(body, request.Passthrough), request.Extensions.KeyOrder)
}
